### ISSUE TITLE
Returning empty array with Promise stuck as Pending

### ISSUE THREAD

**Issue by 97WaterPolo** [JimmyLaurent/torrent-search-api#88]
Hi everyone,

I am currently trying to use torrent-search-api to query and download torrents from IPTorrents. The code I had was working for months without any issue but I had to step away and let it sit for about 3 months. I come back and start it up only to find out none of the queries are working. I haven't touched anything since then. I pulled out only the code relating to the torrent search and set it in a file and I am still running into the same issue.

When I do the query VIA my original code, I am being returned an empty array of torrents. When I stripped out the code and only ran the part that was querying torrent, I was getting an undefined object and it seemed that the promise was stuck in a pending state even using the await function.

I have tried:

- Chaning the URL for IpTorrents to other countries and with/without SSL (I am now running a fresh install of torrent-search-api)
- I am 100% sure my credentials are proper VIA my username and password.
- Adding loops and trying to force the promise to resolve

I have no idea how to catch the errors for the TorrentSearchAPI and am hoping I can find a way to obtain that so I can see what is wrong.

```

const torrentSearch = require('torrent-search-api');
torrentSearch.enableProvider('IpTorrents', 'USER', 'PASSWORD');
const results = startTorrentQuery();
//torrentSearch.enablePublicProviders();
console.log(results);

async function startTorrentQuery() {
    return await torrentSearch.search('1080', 'Movies', 20);
}
```
![badtorrentsearch](https://user-images.githubusercontent.com/6467276/66556803-b7a63e00-eb05-11e9-8d2a-b61f36637064.PNG)

**Comment by JimmyLaurent** [comment]
Try this:
```js

const torrentSearch = require('torrent-search-api');
torrentSearch.enableProvider('IpTorrents', 'USER', 'PASSWORD');

async function startTorrentQuery() {
    return await torrentSearch.search('1080', 'Movies', 20);
}

(async () => {
 const results = await startTorrentQuery();
 console.log(results);
})();
```

**Comment by 97WaterPolo** [comment]
Ran that, but now I am getting an empty array when printing out the value of the search. Copy pasted your code and all I have are the empty square brackets. I know for sure my credentials and I know there should be results for 1080 in the search option.

**Comment by 97WaterPolo** [comment]
> 
> 
> Try this:
> 
> ```js
> const torrentSearch = require('torrent-search-api');
> torrentSearch.enableProvider('IpTorrents', 'USER', 'PASSWORD');
> 
> async function startTorrentQuery() {
>     return await torrentSearch.search('1080', 'Movies', 20);
> }
> 
> (async () => {
>  const results = await startTorrentQuery();
>  console.log(results);
> })();
> ```

Ran that, but now I am getting an empty array when printing out the value of the search. Copy pasted your code and all I have are the empty square brackets. I know for sure my credentials and I know there should be results for 1080 in the search option.

I have tried a few other public providers and it seems to be throwing some random errors occasionally when searching for even basic torrents. Specifically whenever I deal with IPtorrents, I receive an empty array when executing my torrent search, did something change on their end?

**Comment by JimmyLaurent** [comment]
It seems so, unfortunately I don't have an IpTorrents account anymore.

From what I see, your queries may be blocked by cloudfare. 

You can try to enable the "enableCloudFareBypass" flag on the IpTorrents provider.

```json
{
    "name": "IpTorrents",
    "baseUrl": "https://iptorrents.eu",
    "enableCloudFareBypass": true, // <==== Here
    "requireAuthentification": true,
    "supportCookiesAuthentification": true,
    "supportCredentialsAuthentification": true,
    "loginUrl": "/take_login.php",
    "loginQueryString": "username={username}&password={password}",
    "searchUrl": "/t?{cat}q={query};o=seeders",
    "categories": {
       "All": "",
       "Movies": "7;68;54;77;20;89;90;96;6;48;62;38;",
       "TV": "73;",
       "Games": "74;",
       "Music": "75;"
    },
    "defaultCategory": "All",
    "resultsPerPageCount": 50,
    "itemsSelector": "table[id=\"torrents\"] tr:not(:first-child)",
    "itemSelectors": {
       "title": "td:nth-child(2) a",
       "time": "tr:nth-child(2) div.t_ctime | split:%SPECIAL_CHAR%,1 | trim",
       "downloaded": "td:nth-child(7) | int",
       "seeds": "td:nth-child(8) | int",
       "peers": "td:nth-child(9) | int",
       "size": "td:nth-child(6)",
       "link": "td:nth-child(4) a@href",
       "desc": "td:nth-child(2) a@href"
    },
    "paginateSelector": "a:contains(Next)@href",
    "torrentDetailsSelector": ".desWrap@html"
 }
```

**Comment by 97WaterPolo** [comment]
> 
> 
> It seems so, unfortunately I don't have an IpTorrents account anymore.
> 
> From what I see, your queries may be blocked by cloudfare.
> 
> You can try to enable the "enableCloudFareBypass" flag on the IpTorrents provider.
> 
> ```json
> {
>     "name": "IpTorrents",
>     "baseUrl": "https://iptorrents.eu",
>     "enableCloudFareBypass": true, // <==== Here
>     "requireAuthentification": true,
>     "supportCookiesAuthentification": true,
>     "supportCredentialsAuthentification": true,
>     "loginUrl": "/take_login.php",
>     "loginQueryString": "username={username}&password={password}",
>     "searchUrl": "/t?{cat}q={query};o=seeders",
>     "categories": {
>        "All": "",
>        "Movies": "7;68;54;77;20;89;90;96;6;48;62;38;",
>        "TV": "73;",
>        "Games": "74;",
>        "Music": "75;"
>     },
>     "defaultCategory": "All",
>     "resultsPerPageCount": 50,
>     "itemsSelector": "table[id=\"torrents\"] tr:not(:first-child)",
>     "itemSelectors": {
>        "title": "td:nth-child(2) a",
>        "time": "tr:nth-child(2) div.t_ctime | split:%SPECIAL_CHAR%,1 | trim",
>        "downloaded": "td:nth-child(7) | int",
>        "seeds": "td:nth-child(8) | int",
>        "peers": "td:nth-child(9) | int",
>        "size": "td:nth-child(6)",
>        "link": "td:nth-child(4) a@href",
>        "desc": "td:nth-child(2) a@href"
>     },
>     "paginateSelector": "a:contains(Next)@href",
>     "torrentDetailsSelector": ".desWrap@html"
>  }
> ```

I just attempted again with cloud flare bypass, but still thoriwng error, this time with the CloudFlare technique changing.
![plexconsoel](https://user-images.githubusercontent.com/6467276/66737595-432f1000-ee21-11e9-8a7b-9f745ddb106e.PNG)

**Comment by JimmyLaurent** [comment]
The library we use to bypass cloudfare (cloudscraper) may be outdated.
We use the 3.2.0 and the latest is 4.3.0.

You should try to upgrade, but you may have some breaking changes to fix.

**Comment by JimmyLaurent** [comment]
Duplicate of #68

**Comment by 97WaterPolo** [comment]
As of today I started getting this issue again. I have had 	"enableCloudFareBypass": true, in my iptorrents.json and now I am once again recieving the error in regards to Cloudflare changing their technique. I updated to the latest version of cloudscraper(4.5.0) and am still having this issue. If I set the variable to false I am given an empty array of torrent each time the search is executed.

Any idea on how I should resolve this? Only happens with IpTorrents, error and configuration below.

`{
    "name": "IpTorrents",
    "baseUrl": "https://iptorrents.com",
	"enableCloudFareBypass": true,
    "requireAuthentification": true,
    "supportCookiesAuthentification": true,
    "supportCredentialsAuthentification": true,
    "loginUrl": "/take_login.php",
    "loginQueryString": "username={username}&password={password}",
    "searchUrl": "/t?{cat}q={query};o=seeders",
    "categories": {
       "All": "",
       "Movies": "7;68;54;77;20;89;90;96;6;48;62;38;",
       "TV": "73;",
       "Games": "74;",
       "Music": "75;"
    },
    "defaultCategory": "All",
    "resultsPerPageCount": 50,
    "itemsSelector": "table[id=\"torrents\"] tr:not(:first-child)",
    "itemSelectors": {
       "title": "td:nth-child(2) a",
       "time": "tr:nth-child(2) div.t_ctime | split:%SPECIAL_CHAR%,1 | trim",
       "downloaded": "td:nth-child(7) | int",
       "seeds": "td:nth-child(8) | int",
       "peers": "td:nth-child(9) | int",
       "size": "td:nth-child(6)",
       "link": "td:nth-child(4) a@href",
       "desc": "td:nth-child(2) a@href"
    },
    "paginateSelector": "a:contains(Next)@href",
    "torrentDetailsSelector": ".desWrap@html"
 }`

`(node:8372) UnhandledPromiseRejectionWarning: ParserError:
### Cloudflare may have changed their technique, or there may be a bug.
### Bug Reports: https://github.com/codemanki/cloudscraper/issues
### Check the detailed exception message that follows for the cause.

SyntaxError: Challenge evaluation failed: Unexpected token |
    at solveChallenge (C:\Users\Plex\Desktop\TorrentTalker\node_modules\torrent-
search-api\node_modules\cloudscraper\index.js:254:21)
    at Timeout._onTimeout (C:\Users\Plex\Desktop\TorrentTalker\node_modules\torr
ent-search-api\node_modules\cloudscraper\index.js:179:7)
    at ontimeout (timers.js:436:11)
    at tryOnTimeout (timers.js:300:5)
    at listOnTimeout (timers.js:263:5)
    at Timer.processTimers (timers.js:223:10)
(node:8372) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This
error originated either by throwing inside of an async function without a catch
block, or by rejecting a promise which was not handled with .catch(). (rejection
 id: 2)
`

**Comment by JimmyLaurent** [comment]
Cloudfare may have done some changes since someone also opened an issue on cloudscraper with the same problem few hours ago ([issue](https://github.com/codemanki/cloudscraper/issues/325)).

Hopefully, there will be a fix and we'll just have to upgrade.

**Comment by 97WaterPolo** [comment]
A new update was pushed for cloudscraper(version 4.6.0), but still having the same issue with the request being rejected. Any other ideas?
